File: python/examples/connector_template.py
# SPDX-License-Identifier: Apache-2.0
"""Connector template — your starting point (Python).

Copy this file, edit the ONE function marked `EDIT` below, and you have a
working connector. Everything else is done for you.

See a sealed event right now — no key, no NATS, no feed:

    echo '{"lat":26.4,"lon":50.9,"alt_m":11000,"quality":0.9}' \\
      | python examples/connector_template.py --dry-run

Then run for real (key from scripts/gen-connector-key.sh):

    AJAR_SIGNING_SEED=connector.seed AJAR_SOURCE_ID=acme-radar-1 \\
    NATS_URL=nats://nats.you.mil:4222  python examples/connector_template.py

Production behaviour (built in, no edits needed):
- A malformed or un-mappable record is logged and skipped, never fatal.
- A NATS publish error is logged and the connector keeps going.
- Set AJAR_HEALTH_ADDR=0.0.0.0:9090 to expose GET /healthz (liveness) and
  GET /metrics (Prometheus counters: published / skipped / publish_errors).
"""

import asyncio
import json
import os
import sys
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer

from ajar_connector import EventBuilder, SigningKey, canonical_bytes, seal
from ajar_connector.event_pb2 import Event

# Process counters, surfaced at /metrics.
_METRICS = {"published": 0, "skipped": 0, "publish_errors": 0}


# ┌───────────────────────────────────────────────────────────────────────────┐
# │ EDIT — map one record from your feed into a canonical Event.               │
# │ `r` is one JSON object from your feed (change the fields to match yours).  │
# │ Use the entity_type Ajar assigned you. Add .attribute(k, v) only for       │
# │ attributes your entity type's ontology schema defines.                     │
# └───────────────────────────────────────────────────────────────────────────┘
def to_event(source_id: str, r: dict) -> Event:
    return (
        EventBuilder(source_id, "mim:aircraft")
        .new_id()
        .now()
        .location(r["lat"], r["lon"], r["alt_m"])
        .confidence(r["quality"])
        .build()
    )


# ─────────────────────────────────────────────────────────────────────────────
# You usually don't need to touch anything below this line.
# ─────────────────────────────────────────────────────────────────────────────


def _spawn_health() -> None:
    """Start a tiny health/metrics HTTP server if AJAR_HEALTH_ADDR is set
    (e.g. 0.0.0.0:9090). GET /healthz -> liveness; GET /metrics -> Prometheus
    text. Stdlib only — no impact unless you opt in."""
    addr = os.environ.get("AJAR_HEALTH_ADDR")
    if not addr:
        return
    host, _, port = addr.rpartition(":")

    class _Handler(BaseHTTPRequestHandler):
        def do_GET(self):  # noqa: N802
            if self.path.startswith("/metrics"):
                body = (
                    "# TYPE ajar_connector_published_total counter\n"
                    f"ajar_connector_published_total {_METRICS['published']}\n"
                    "# TYPE ajar_connector_skipped_total counter\n"
                    f"ajar_connector_skipped_total {_METRICS['skipped']}\n"
                    "# TYPE ajar_connector_publish_errors_total counter\n"
                    f"ajar_connector_publish_errors_total {_METRICS['publish_errors']}\n"
                )
            else:
                body = "ok\n"
            data = body.encode()
            self.send_response(200)
            self.send_header("Content-Type", "text/plain")
            self.send_header("Content-Length", str(len(data)))
            self.end_headers()
            self.wfile.write(data)

        def log_message(self, *_args):  # silence per-request logging
            pass

    server = HTTPServer((host or "0.0.0.0", int(port)), _Handler)
    print(f"[connector] health/metrics on http://{addr}/healthz and /metrics", file=sys.stderr)
    threading.Thread(target=server.serve_forever, daemon=True).start()


def _nats_tls_opts() -> dict:
    """mTLS when AJAR_TLS_CA / AJAR_TLS_CERT / AJAR_TLS_KEY are all set (the
    production path — client-cert CN = source_id, mounted by the Helm chart under
    /etc/ajar/tls). Unset -> plaintext for local dev."""
    ca = os.environ.get("AJAR_TLS_CA")
    cert = os.environ.get("AJAR_TLS_CERT")
    key = os.environ.get("AJAR_TLS_KEY")
    if ca and cert and key:
        import ssl

        ctx = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH, cafile=ca)
        ctx.load_cert_chain(certfile=cert, keyfile=key)
        print("[connector] mTLS enabled (client cert = source identity)", file=sys.stderr)
        return {"tls": ctx}
    print("[connector] no AJAR_TLS_* set — connecting without TLS (dev only)", file=sys.stderr)
    return {}


def _load_seed(dry_run: bool) -> bytes:
    path = os.environ.get("AJAR_SIGNING_SEED")
    if path:
        seed = open(path, "rb").read()
        if len(seed) != 32:
            raise SystemExit("AJAR_SIGNING_SEED file must be exactly 32 bytes")
        return seed
    if dry_run:
        print("[connector] no AJAR_SIGNING_SEED set — ephemeral throwaway key (dry-run only)", file=sys.stderr)
        return os.urandom(32)
    raise SystemExit("set AJAR_SIGNING_SEED to your 32-byte key file (see scripts/gen-connector-key.sh)")


async def main() -> None:
    dry_run = "--dry-run" in sys.argv
    source_id = os.environ.get("AJAR_SOURCE_ID", "demo-connector")
    prefix = os.environ.get("AJAR_INGEST_PREFIX", "ajar.ingest")
    nats_url = os.environ.get("NATS_URL", "nats://127.0.0.1:4222")
    subject = f"{prefix}.{source_id}"
    key = SigningKey.from_seed(_load_seed(dry_run))

    _spawn_health()  # no-op unless AJAR_HEALTH_ADDR is set

    nc = None
    if dry_run:
        print("[connector] --dry-run: building + sealing, not publishing", file=sys.stderr)
    else:
        import nats  # transport dep — only needed for a real run

        print(f"[connector] connecting to NATS at {nats_url}", file=sys.stderr)
        nc = await nats.connect(nats_url, **_nats_tls_opts())
    print(f"[connector] source_id={source_id}  subject={subject}", file=sys.stderr)

    # Your feed: by default, newline-delimited JSON on stdin. Swap this loop for
    # your TCP socket / file / API / serial port — the rest stays the same.
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        # Resilient: a bad record is logged and skipped, never fatal.
        try:
            record = json.loads(line)  # parse one record
            event = to_event(source_id, record)  # EDIT maps it
        except Exception as e:  # noqa: BLE001 — one bad input must not kill the loop
            print(f"[connector] skip: bad record: {e}", file=sys.stderr)
            _METRICS["skipped"] += 1
            continue
        canonical = canonical_bytes(event)
        sealed = seal(canonical, key)  # sign it
        if nc is not None:
            try:
                await nc.publish(subject, sealed)  # publish it
            except Exception as e:  # noqa: BLE001 — keep running on a transport blip
                print(f"[connector] publish error (continuing): {e}", file=sys.stderr)
                _METRICS["publish_errors"] += 1
                continue
        _METRICS["published"] += 1
        tag = "" if nc is not None else "  [dry-run]"
        print(f"{event.id} -> {subject} ({len(sealed)} sealed bytes){tag}")

    if nc is not None:
        await nc.flush()
        await nc.close()


if __name__ == "__main__":
    asyncio.run(main())

File: cpp/examples/connector_template.cpp
// SPDX-License-Identifier: Apache-2.0
//
// Connector template (C++): a copy-me starting point for a new Ajar connector.
// Make the two edits marked EDIT 1 and EDIT 2 below, then build and run.
//
// Build (from the repo root):
//   cmake -S cpp -B cpp/build -DCMAKE_BUILD_TYPE=Release && cmake --build cpp/build -j
//
// See a sealed event right now — no key, no NATS, no feed. The demo input is one
// record per line, "lat lon alt_m quality":
//   echo "26.4 50.9 11000 0.9" | ./cpp/build/connector_template --dry-run
//
// Run for real (key from scripts/gen-connector-key.sh; mTLS materials + endpoint
// issued by your operator). Publishing needs the cnats client, auto-detected by
// CMake; without it the binary still builds and runs in --dry-run:
//   AJAR_TLS_CA=ca.pem AJAR_TLS_CERT=client.pem AJAR_TLS_KEY=client.key \
//   AJAR_SIGNING_SEED=connector.seed AJAR_SOURCE_ID=acme-radar-1 \
//   NATS_URL=tls://nats.you.mil:443  ./cpp/build/connector_template

#include <array>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

#include "ajar/connector.hpp"
#ifdef AJAR_WITH_NATS
#include <nats/nats.h>
#endif

namespace {

// ╔═══════════════════════════════════════════════════════════════════════════╗
// ║ EDIT 1 — describe ONE record from your feed, and how to parse one line.   ║
// ║ The demo reads "lat lon alt_m quality"; swap in YOUR feed's format.       ║
// ╚═══════════════════════════════════════════════════════════════════════════╝
struct MyRecord {
  double lat = 0, lon = 0, alt_m = 0, quality = 0;
};

bool parse_record(const std::string& line, MyRecord& out) {
  std::istringstream is(line);
  return static_cast<bool>(is >> out.lat >> out.lon >> out.alt_m >> out.quality);
}

// ╔═══════════════════════════════════════════════════════════════════════════╗
// ║ EDIT 2 — map your record into a canonical Event.                          ║
// ║ Use the entity_type your operator assigned. Add .attribute(k, v) only for ║
// ║ attributes that type's ontology schema defines.                          ║
// ╚═══════════════════════════════════════════════════════════════════════════╝
ajar::Event to_event(const std::string& source_id, const MyRecord& r) {
  return ajar::EventBuilder(source_id, "mim:aircraft")
      .new_id()
      .now()
      .location(r.lat, r.lon, r.alt_m)
      .confidence(r.quality)
      .build();
}

// ─────────────────────────────────────────────────────────────────────────────
// You usually don't need to touch anything below this line.
// ─────────────────────────────────────────────────────────────────────────────

std::string env_or(const char* key, const char* def) {
  const char* v = std::getenv(key);
  return (v && *v) ? std::string(v) : std::string(def);
}

// Load the 32-byte seed from the file named by AJAR_SIGNING_SEED; in --dry-run
// with none set, fall back to a dev seed (never used for real publishing).
std::array<std::uint8_t, 32> load_seed(bool dry_run) {
  const char* path = std::getenv("AJAR_SIGNING_SEED");
  if (path && *path) {
    std::ifstream f(path, std::ios::binary);
    std::array<std::uint8_t, 32> seed{};
    if (!f.read(reinterpret_cast<char*>(seed.data()), seed.size()) || f.gcount() != 32)
      throw std::runtime_error("signing seed file must be exactly 32 bytes");
    return seed;
  }
  if (dry_run) {
    std::fprintf(stderr, "[connector] no AJAR_SIGNING_SEED set — using a DEV seed (dry-run only)\n");
    std::array<std::uint8_t, 32> seed{};
    seed.fill(0x03);
    return seed;
  }
  throw std::runtime_error("set AJAR_SIGNING_SEED to your 32-byte seed file (see scripts/gen-connector-key.sh)");
}

}  // namespace

int main(int argc, char** argv) {
  bool dry_run = false;
  bool check_only = false;
  for (int i = 1; i < argc; ++i) {
    if (std::strcmp(argv[i], "--dry-run") == 0) dry_run = true;
    if (std::strcmp(argv[i], "--check") == 0) check_only = true;
  }

  // --check validates what EDIT 2 actually builds against the vendored
  // ontology, then exits — no key, no feed, no network. Run it in CI so a
  // mapping mistake fails the build instead of publishing events Ajar
  // silently discards.
  if (check_only) {
    const ajar::Event probe = to_event("check", MyRecord{});
    const auto faults = ajar::validate(probe);
    for (const auto& f : faults)
      std::fprintf(stderr, "[check] %s\n", f.message().c_str());
    std::fprintf(stderr, "[check] %s against ontology %s\n",
                 faults.empty() ? "mapping is clean" : "mapping has faults",
                 ajar::ontology_version());
    return faults.empty() ? 0 : 1;
  }

  const std::string source_id = env_or("AJAR_SOURCE_ID", "demo-connector");
  const std::string prefix = env_or("AJAR_INGEST_PREFIX", "ajar.ingest");
  const std::string nats_url = env_or("NATS_URL", "nats://127.0.0.1:4222");
  const std::string subject = prefix + "." + source_id;

  ajar::SigningKey key = ajar::SigningKey::from_seed(load_seed(dry_run));

#ifdef AJAR_WITH_NATS
  natsConnection* conn = nullptr;
  if (!dry_run) {
    std::fprintf(stderr, "[connector] connecting to NATS at %s\n", nats_url.c_str());
    natsOptions* opts = nullptr;
    natsOptions_Create(&opts);
    natsOptions_SetURL(opts, nats_url.c_str());
    // mTLS when AJAR_TLS_CA/CERT/KEY are all set; plaintext for local dev.
    const char* ca = std::getenv("AJAR_TLS_CA");
    const char* cert = std::getenv("AJAR_TLS_CERT");
    const char* keyf = std::getenv("AJAR_TLS_KEY");
    if (ca && *ca && cert && *cert && keyf && *keyf) {
      std::fprintf(stderr, "[connector] mTLS enabled (client cert = source identity)\n");
      natsOptions_SetSecure(opts, true);
      natsOptions_LoadCATrustedCertificates(opts, ca);
      natsOptions_LoadCertificatesChain(opts, cert, keyf);
    } else {
      std::fprintf(stderr, "[connector] no AJAR_TLS_* set — connecting without TLS (dev only)\n");
    }
    natsStatus st = natsConnection_Connect(&conn, opts);
    natsOptions_Destroy(opts);
    if (st != NATS_OK) {
      std::fprintf(stderr, "connect NATS: %s\n", natsStatus_GetText(st));
      return 1;
    }
  }
#else
  if (!dry_run) {
    std::fprintf(stderr, "built without the NATS client — rebuild with cnats to publish, or use --dry-run\n");
    return 1;
  }
#endif

  std::fprintf(stderr, "[connector] source_id=%s  subject=%s%s\n", source_id.c_str(),
               subject.c_str(), dry_run ? "  [dry-run]" : "");

  // Your feed: by default, one record per line on stdin ("lat lon alt_m quality").
  // Swap this loop for your socket / file / API / serial port — the rest stays.
  std::string line;
  while (std::getline(std::cin, line)) {
    if (line.empty()) continue;
    // Resilient: a bad record is logged and skipped, never fatal.
    MyRecord rec;
    if (!parse_record(line, rec)) {
      std::fprintf(stderr, "[connector] skip: malformed record: %s\n", line.c_str());
      continue;
    }
    ajar::Event event;
    try {
      event = to_event(source_id, rec);
    } catch (const std::exception& ex) {
      std::fprintf(stderr, "[connector] skip: cannot map record: %s\n", ex.what());
      continue;
    }
    const std::string canonical = ajar::canonical_bytes(event);
    const std::vector<std::uint8_t> sealed = ajar::seal(canonical, key);
#ifdef AJAR_WITH_NATS
    if (conn) {
      natsStatus st = natsConnection_Publish(conn, subject.c_str(), sealed.data(),
                                             static_cast<int>(sealed.size()));
      if (st != NATS_OK) {
        std::fprintf(stderr, "[connector] publish error (continuing): %s\n", natsStatus_GetText(st));
        continue;
      }
    }
#endif
    std::printf("%s -> %s (%zu sealed bytes)%s\n", event.id().c_str(), subject.c_str(),
                sealed.size(), dry_run ? "  [dry-run]" : "");
  }

#ifdef AJAR_WITH_NATS
  if (conn) {
    natsConnection_Flush(conn);
    natsConnection_Destroy(conn);
  }
#endif
  return 0;
}
